doc/api/esm.md

```diff
@@ -1096,20 +1099,187 @@ export async function resolve(specifier,
 With this loader, running:
 
 ```console
-NODE_OPTIONS='--experimental-loader ./custom-loader.mjs' node x.js
+NODE_OPTIONS='--experimental-loader ./custom-loader.mjs' node main.js
 ```
 
-would load the module `x.js` as an ES module with relative resolution support
+would load the module `main.js` as an ES module with relative resolution support
 (with `node_modules` loading skipped in this example).
 
+### Get format hook
+
+> Note: The loaders API is being redesigned. This hook may disappear or its
+> signature may change. Do not rely on the API described below.
+
+The `getFormat` hook provides a way to define a custom method of determining how
+a URL should be interpreted. Using this hook will cause Node.js to ignore any
+format returned from `resolve`.
+
+```js
+import { request } from 'https';
+
+const formatMap = {
+  'text/javascript': 'module',
+  'application/javascript': 'module'
+};
+/**
+ * @param {{url}} parameters
+ * @returns {{format}}
+ */
+export async function getFormat({ url, isMain }, defaultGetFormat) {
+  if (url.startsWith('https://')) {
+    return new Promise((fulfill, reject) => {
+      request(url, {
+        method: 'HEAD'
+      }, (err, res) => {
+        if (err) reject(err);
+        else fulfill(formatMap[res.headers['content-type']]);
+      });
+    });
+  }
+  return defaultGetFormat({ url, isMain });
```

Comment:
Per additional props, doing so requires no changes to the code currently. I'm not willing to document an unknown set of properties being spread if thats the question as that likely just will lead to confusion.

Per `undefined`,  hooks must delegate to the default. Returning a value such as `undefined` allows them to censor or prevent default behavior. They can and should use `defaultGetFormat` to get default behavior.

Reply:
Right now `defaultGetFormat` requires `url` and `isMain` properties.  My suggestion of `defaultGetFormatFactory` is to deal with the possibility of it needing an additional property in the future.  I can see how object spread is not the best way to deal with this and the following would be better?
```js
function defaultGetFormatFactory(defaultProps) {
  return ({
    url = defaultProps.url,
    isMain = defaultProps.isMain
  } = {}) => defaultGetFormat({ url, isMain });
}
```
More verbose factory function but avoids spread of potentially unknown properties and allows `getFormat` hooks to simply call `defaultGetFormat()`.  So if `defaultGetFormat` needs a patch where it requires a `packageType` property in the future that could be non-breaking to existing hooks.

Wouldn't `{format: undefined}` prevent default behavior (object with undefined format property rather than `undefined`)?  If a `getFormat` hook directly returns `undefined` it will currently throw `TypeError: Cannot destructure property 'format' of ...`.

Reply:
I don't understand how this changes the example code. We would add more properties in the example as they are added to the codebase.

Reply:
Adding properties to the example code won't fix existing loader hooks, if someone writes a loader hook which calls `defaultGetFormat({ url, isMain })` and a future version of node.js requires an additional property, that loader hook will break.

Reply:
@coreyfarrell correct, they will need to update their loader. We do not know all possible new fields that loaders may be using are guaranteed to be safe to delegate / have no effect.

Reply:
Agreed this is an important concern for upgrade paths. If the point of using an object is for upgrade paths, then it is very very easy for a loader to be incompatible with such upgrade paths using this suggested pattern.

Personally I find this a good reason not to have the first argument be an object, and rather just have it as the string URL. `isMain` is no longer supported so perhaps we can just leave it out entirely, since we likely don't want to select on that anyway.

Reply:
Suggestion:

```js
getFormat (url: String, defaultGetFormat)
```